marshal/cluster.go

```diff
@@ -0,0 +1,272 @@
+/*
+ * portal - marshal
+ *
+ * a library that implements an algorithm for doing consumer coordination within Kafka, rather
+ * than using Zookeeper or another external system.
+ *
+ */
+
+package marshal
+
+import (
+	"crypto/md5"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/dropbox/kafka"
+)
+
+var (
+	clusters    = make(map[string]*cluster)
+	clusterLock = &sync.Mutex{}
+)
+
+// cluster is a user-agnostic view of the world. It connects to a Kafka cluster
+// and runs rationalizers to observe the complete world state.
+type cluster struct {
+	// These members are not protected by the lock and can be read at any
+	// time as they're write-once or only ever atomically updated. They must
+	// never be overwritten once a cluster is created.
+	quit       *int32
+	broker     *kafka.Broker
+	producer   kafka.Producer
+	partitions int
+	jitters    chan time.Duration
+
+	// Lock protects the following members; you must have this lock in order to
+	// read from or write to these.
+	lock   sync.RWMutex
+	topics map[string]int
+	groups map[string]map[string]*topicState
+
+	// This WaitGroup is used for signalling when all of the rationalizers have
+	// finished processing.
+	rationalizers sync.WaitGroup
+
+	// rsteps is updated whenever a rationalizer processes a log entry, this is
+	// used mainly by the test suite.
+	rsteps *int32
+
+	// This is for testing only. When this is non-zero, the rationalizer will answer
+	// queries based on THIS time instead of the current, actual time.
+	ts int64
+}
+
+// getOrCreateCluster will either return a pointer to a valid cluster object if it exists or
+// create a new one and then return it.
+func getOrCreateCluster(brokers []string) (*cluster, error) {
+	sort.Strings(brokers)
+	clusterKey := strings.Join(brokers, ",")
```

Comment:
I had a concern about this and how it will interact with service discovery. I think it will be okay but consider:

Suppose the JQS calls `NewMarshaler` at two different times, both logically for the same Kafka cluster, but for whatever reason one of the brokers is missing in the 2nd call. This will result in us creating a new `cluster` object here that in fact points to the same Kafka as an existing one. Both of these belong to Marshallers with the same clientId, and groupId can also be identical, so in theory they could conflict.

However, the JQS creates at most one `dropbox/kafka.Consumer` per `jqs.kafkaConsumerKey`, so even though these two local Marshallers look identical to the Marshal protocol, they will never be trying to claim the same topic at the same time. So they will in fact not conflict.